modules/um_regonlymsg.c

```diff
@@ -0,0 +1,135 @@
+/*
+ * modules/um_regonlymsg.c
+ * Copyright (c) 2020 Ariadne Conill <ariadne@dereferenced.org>
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice,
+ *    this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ * 3. The name of the author may not be used to endorse or promote products
+ *    derived from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
+ * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT,
+ * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+ * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
+ * IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "stdinc.h"
+#include "modules.h"
+#include "hook.h"
+#include "client.h"
+#include "ircd.h"
+#include "send.h"
+#include "hash.h"
+#include "s_conf.h"
+#include "s_user.h"
+#include "s_serv.h"
+#include "numeric.h"
+#include "privilege.h"
+#include "s_newconf.h"
+
+static int
+um_regonlymsg_modinit(void)
+{
+	user_modes['R'] = find_umode_slot();
+	construct_umodebuf();
+
+	return 0;
+}
+
+static void
+um_regonlymsg_moddeinit(void)
+{
+	user_modes['R'] = 0;
+	construct_umodebuf();
+}
+
+#define IsSetRegOnlyMsg(c)	((c->umodes & user_modes['R']) == user_modes['R'])
+
+static const char um_regonlymsg_desc[] =
+	"Provides usermode +R which restricts messages from unregistered users.";
+
+static bool
+allow_message(struct Client *source_p, struct Client *target_p)
+{
+	if (!IsSetRegOnlyMsg(target_p))
+		return true;
+
+	if (IsServer(source_p))
+		return true;
+
+	/* XXX: controversial?  allow opers to send through +R */
+	if (IsOper(source_p))
+		return true;
+
+	if (accept_message(source_p, target_p))
```

Comment:
This will be called on the wrong server in the invite case.

Reply:
i don't follow.  if you check the diff, the order is the same verses what was removed.

Reply:
On a multi-server network, does this not result in calling `accept_message` on `source_p`'s server, and thus dereferencing a NULL `target_p->localClient`?

Reply:
Ah, I see what you mean.  Because the hook is called in MyClient(source_p) context.  Well, that is something that can be overcome.

Reply:
I reworked the way the hook is called in m_invite.  Seems to be working safely now.

Reply:
Would it not have been easier to change it to use the `invite` hook (called in `MyClient(target_p)` context) instead of the `can_invite` hook? Then you can revert 4371dcb entirely and revert this line change https://github.com/charybdis-ircd/charybdis/pull/341/files#diff-890476bb23a79380cafafdde59c6a339R161 too.

Reply:
That's a good idea.  I'll go ahead and do that.

Reply:
Done.